pennylane/liealg/cartan_decomp.py

```diff
@@ -0,0 +1,305 @@
+    >>> ops1 = [qml.X(0)]
+    >>> ops2 = [qml.Y(0)]
+    >>> vspace1 = [qml.X(0), qml.Y(0)]
+
+    Because :math:`[X_0, Y_0] = 2i Z_0`, the commutators do not map to the selected vector space.
+
+    >>> check_commutation(ops1, ops2, vspace1)
+    False
+
+    Instead, we need the full :math:`\mathfrak{su}(2)` space.
+
+    >>> vspace2 = [qml.X(0), qml.Y(0), qml.Z(0)]
+    >>> check_commutation(ops1, ops2, vspace2)
+    True
+    """
+
+    ops1_is_tensor = any(isinstance(op, TensorLike) for op in ops1)
+    ops2_is_tensor = any(isinstance(op, TensorLike) for op in ops2)
+    if not isinstance(vspace, qml.pauli.PauliVSpace):
+        vspace_is_tensor = any(isinstance(op, TensorLike) for op in vspace)
+        if any(isinstance(op, Operator) for op in vspace):
+            vspace = qml.pauli.PauliVSpace(vspace, dtype=complex)
+
+    else:
+        vspace_is_tensor = False
+
+    if not all((ops1_is_tensor, ops2_is_tensor, vspace_is_tensor)) and any(
+        (ops1_is_tensor, ops2_is_tensor, vspace_is_tensor)
+    ):
+        raise TypeError(
+            "All inputs ``ops1``, ``ops2`` and ``vspace`` to qml.liealg.check_commutation need to either be iterables of operators or matrices."
+        )
+
+    if all((ops1_is_tensor, ops2_is_tensor, vspace_is_tensor)):
+        return _check_commutation_matrix(ops1, ops2, vspace)
+
+    if any(isinstance(op, Operator) for op in ops1):
+        ops1 = [op.pauli_rep for op in ops1]
+
+    if any(isinstance(op, Operator) for op in ops2):
+        ops2 = [op.pauli_rep for op in ops2]
+
+    for o1 in ops1:
+        for o2 in ops2:
+            com = o1.commutator(o2)
+            com.simplify()
+            if len(com) != 0:
+                if vspace.is_independent(com):
+                    return False
+
+    return True
+
+
+def _all_coms(vspace1, vspace2):
+    r"""Compute all commutators [Vspace1, Vspace2]"""
+    chi = len(vspace1[0])
+
+    m0m1 = qml.math.moveaxis(qml.math.tensordot(vspace1, vspace2, axes=[[2], [1]]), 1, 2)
+    m0m1 = qml.math.reshape(m0m1, (-1, chi, chi))
+
+    # Implement einsum "aij,bki->abkj" by tensordot and moveaxis
+    m1m0 = qml.math.moveaxis(qml.math.tensordot(vspace1, vspace2, axes=[[1], [2]]), 1, 3)
+    m1m0 = qml.math.reshape(m1m0, (-1, chi, chi))
+    all_coms = m0m1 - m1m0
+    return all_coms
+
+
+def _is_subspace(subspace, vspace):
+    r"""check if subspace <= vspace"""
+    # Check if rank increases by adding matices from ``subspace``
+    # Use matrices as vectors -> flatten matrix dimensions (chi, chi) to (chi**2,)
+    vspace = qml.math.reshape(vspace, (len(vspace), -1))
+    subspace = qml.math.reshape(subspace, (len(subspace), -1))
+
+    rank_V = qml.math.linalg.matrix_rank(vspace)
+    rank_both = qml.math.linalg.matrix_rank(qml.math.vstack([vspace, subspace]))
+
+    if rank_both > rank_V:
+        return False
+    return True
+
+
+def _check_commutation_matrix(a, b, vspace):
+    all_coms = _all_coms(a, b)
+    return _is_subspace(all_coms, vspace)
```

Comment:
Maybe not needed to put this in a separate function? Non-blocking question for sure.

Reply:
yes good point! will remove